### ISSUE TITLE
correct callback in custom renderer for repaint/tick

### ISSUE THREAD

**Issue by dakom** [facebook/react#10950]
from some previous discussions, I gather that there is a way to render the whole tree when it's ready (e.g. flushSync / updateContainer) - however, I don't see a way to get a callback when it's completed. Is that possible via one of the builtin functions?

The way I'm dealing with it now is to make sure my root node is a higher order component and then put my callback logic in `componentDidUpdate()`. It seems to work so far, but it would be better if this could be a little closer to the metal (yet still only after the entire tree is definitively reconciled and previous renders finished)

Thanks again!

**Comment by dakom** [comment]
Here I put together a code sample to be more specific - it seems each of these methods is called only once. Neither frame ticks nor state/prop updates cause them to be called again, so there is no place to put `doUpdate()`:

```
export const MyRenderer = (function () {
    let container;

    return {
        render: (element, mountPoint, paintScreen) => {
            if (!container) {
                container = Renderer.createContainer(mountPoint);
            }

            const doUpdate = () => {
                Renderer.updateContainer(element, container);
                paintScreen();
            }

            //Kindof just assigning randomly to one of these for now... eini mini miney mo... flushSync!

            Renderer.batchedUpdates(() => {
                console.log("batched updates...");
            });

            Renderer.unbatchedUpdates(() => {
                console.log("unbatched updates...");
            });

            Renderer.flushSync(() => {
                console.log("flushing...");

                doUpdate();
            });

            Renderer.deferredUpdates(() => {
                console.log("deferred updates...");
            });
        },
    }
})();
```

Called like:

```
MyRenderer.render(<Root />, someHolder, () => { /*imperative drawing stuff*/ });
```

**Comment by gaearon** [comment]
`updateContainer` accepts callback in arguments. You can see it being used in ReactDOM. Hope this helps!

**Comment by dakom** [comment]
Yes- that does help, thanks! 

Followup question - if I'm continually calling `updateContainer` on each frame - does the following approach make sense (based off [ReactDOM](https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L743)) ?

**First call** 
1. createContainer (and store a reference)
2. call `updateContainer` via `unbatchedUpdates` callback

**Subsequent calls:**
1. Call `updateContainer` directly (_not_ via any callbacks like flushSync)

**Comment by gaearon** [comment]
Sounds reasonable to me although I don't fully understand why you need to call it yourself. If expect that you let components call `setState` (and that causes the updates).

**Comment by dakom** [comment]
Cool :)

So I'm doing something a bit off the beaten path... when I've finished it to the extent of a demo is there a good place to bring it up for discussion?

On one foot - I have two React renderers: DOM and Custom/WebGL. All State is processed in a Worker thread such that the flow is basically like:

Input -> Worker -> State -> Elements (sent on frame tick) -> React -> Output 

(Output _probably_ allows pushing async events as Input, but I'm still working that out- let's ignore async events for now)

So it's more like a conventional game loop in a way, where the touch detection and things are done by comparing input events to coordinates in State- but with the added benefit that the output is handled by a declarative rendering of state.

I'm really not sure if this is totally wacky since it's not something I see commonly done - e.g. there's no components (just a constantly renewed createElement tree generated from that single state), and it's also still early on without having a big project to prove it on...

But you were wondering why I need to call `render()` myself... I hope I have a good excuse :)

**Comment by gaearon** [comment]
Interesting. If you get it working I'd love to take a look.
